frontend/js/src/common/brainzplayer/BrainzPlayer.tsx

```diff
@@ -128,126 +110,143 @@ function isListenFromDatasource(
+    appleMusicEnabled &&
+      AppleMusicPlayer.hasPermissions(appleAuth) &&
+      "appleMusic",
+    soundcloudEnabled &&
+      SoundcloudPlayer.hasPermissions(soundcloudAuth) &&
+      "soundcloud",
+    youtubeEnabled && "youtube",
+  ].filter(Boolean) as Array<DataSourceKey>;
+
+  const sortedDataSources = dataSourcesPriority.filter((key) =>
+    enabledDataSources.includes(key)
+  );
 
+  // Refs
+  const spotifyPlayerRef = React.useRef<SpotifyPlayer>(null);
+  const youtubePlayerRef = React.useRef<YoutubePlayer>(null);
+  const soundcloudPlayerRef = React.useRef<SoundcloudPlayer>(null);
+  const appleMusicPlayerRef = React.useRef<AppleMusicPlayer>(null);
+  const dataSourceRefs: Array<React.RefObject<
+    DataSourceTypes
+  >> = React.useMemo(() => {
+    const dataSources: Array<React.RefObject<DataSourceTypes>> = [];
     sortedDataSources.forEach((key) => {
       switch (key) {
         case "spotify":
-          this.dataSources.push(this.spotifyPlayer);
+          dataSources.push(spotifyPlayerRef);
           break;
         case "youtube":
-          this.dataSources.push(this.youtubePlayer);
+          dataSources.push(youtubePlayerRef);
           break;
         case "soundcloud":
-          this.dataSources.push(this.soundcloudPlayer);
+          dataSources.push(soundcloudPlayerRef);
           break;
         case "appleMusic":
-          this.dataSources.push(this.appleMusicPlayer);
+          dataSources.push(appleMusicPlayerRef);
           break;
         default:
         // do nothing
       }
     });
-  }
-
-  componentWillUnMount = () => {
-    window.removeEventListener("storage", this.onLocalStorageEvent);
-    window.removeEventListener("message", this.receiveBrainzPlayerMessage);
-    window.removeEventListener("beforeunload", this.alertBeforeClosingPage);
-    this.stopPlayerStateTimer();
-  };
-
-  alertBeforeClosingPage = (event: BeforeUnloadEvent) => {
-    const { playerPaused } = this.state;
+    return dataSources;
+  }, [sortedDataSources]);
+
+  const playerStateTimerID = React.useRef<NodeJS.Timeout | null>(null);
+  const queueRef = React.useRef<BrainzPlayerQueue>(queue);
+  queueRef.current = queue;
+  const ambientQueueRef = React.useRef<BrainzPlayerQueue>(ambientQueue);
+  ambientQueueRef.current = ambientQueue;
+  const currentListenRef = React.useRef(currentListen);
+  currentListenRef.current = currentListen;
+  const currentDataSourceIndexRef = React.useRef(currentDataSourceIndex);
+  currentDataSourceIndexRef.current = currentDataSourceIndex;
+  const isActivatedRef = React.useRef(isActivated);
+  isActivatedRef.current = isActivated;
+  const currentTrackNameRef = React.useRef(currentTrackName);
+  currentTrackNameRef.current = currentTrackName;
+  const listenSubmittedRef = React.useRef(listenSubmitted);
+  listenSubmittedRef.current = listenSubmitted;
+  const currentListenIndexRef = React.useRef(currentListenIndex);
+  currentListenIndexRef.current = currentListenIndex;
```

Comment:
The refs used here to avoid stale closures. I need to assign the current to these refs, so that whenever the context is updated, the refs have the latest value.

Reply:
Sorry if I'm being thick but I don't understand...

Couldn't you simply use (as an example) `const {ambientQueue} = useBrainzPlayerContext();` directly inside the functions that currently use ambientQueueRef.current (in this case in playNextTrack)?
I get that using useBrainzPlayerContext in the main body of the function and using values inside a closure results in stale values, but I expect using the context inside the functions directly should results in always fresh values.

Reply:
No, we cannot get the context value in the function as `useBrainzPlayerContext` is a hook and Hooks can only be called inside of the body of a function component. So that'll throw an error as that breaks the rules of hooks.